### ISSUE TITLE
Upgrade proftpd from apt-get installed version to compiled version not working

### ISSUE THREAD

**Issue by mdell-seradex** [proftpd/proftpd#1275]
I am sorry for posting this here but since the forums are still unable to register new users and others have done this, I am posting this here.  I appreciate any help you can provide.

### What I Did
On a Ubuntu 20.04 LTS HyperV image I used the following command to originally install proftpd
`sudo apt-get install proftpd`
This installed ProFTPD Version 1.3.6c
Later I discovered I needed additional functionality that is not provided in the default install.
To do this I downloaded 1.3.7a, configured it, compiled it, and installed it.
I have now discovered that I need more functionality.
I have a second virtual machine that is the original copy of the one in production and it has not yet been upgraded to 1.3.7a, yet I cannot seem to do it.  I reviewed the command history in the terminal window of the production machine and cannot find anything that I have missed.  I was trying to build proftpd so it was as close as possible to the version apt-get installed.  Also, I matched the commands I found I used in the production server, but failed to make it work.  One thing I found is that the make install command appears to be installing proftpd into a different location ("/usr/local" instead of "/usr"), and it seems that some files are in different folders as well when comparing the two locations.
These are the commands I used.  Note that I inserted comments to explain with I did some things.  I also grabbed my configuration files from the production server and transferred them to this test server, changing IPs as needed.
```
sudo apt-get update -y
/* This is to install cap support as the apt-get version has it */
sudo apt-get install libcap2-bin
/* do clean up */
sudo apt autoremove
cd /usr/src/
/* extract the 1.3.7a version I had */
sudo tar -xzf proftpd-1.3.7a.tar.gz
/* install the build tools */
sudo apt-get install build-essential
sudo apt-get update -y
sudo apt autoremove
cd proftpd-1.3.7a/
sudo make clean
/* I know some things are supposed to be detected and weren't.  I do not know why, but this worked in the production server */
sudo ./configure --enable-dso --enable-ctrls --with-modules=mod_ident:mod_rewrite
sudo make
/* compare the built modules to the pre-installed modules to confirm that they are close enough */
./proftpd -l
proftpd -l
/* confirm that there are no discernible issues with the build and my configuration */
sudo ./proftpd -t -d10 -c /etc/proftpd/proftpd.conf
/* try to install it */
sudo systemctl stop proftpd
sudo make install
sudo systemctl start proftpd
/* discover there is an issue with it starting */
sudo systemctl status proftpd
/ check the installed version and built version to confirm they are the same - they are not. The first still reports 1.3.6c and the second reports 1.3.7a */
proftpd -v
./proftpd -v
```

### What I Expected/Wanted
What I wanted was for the version that is installed as a service to be updated by the one I built and installed, but it does not appear to be working.  I noticed in the configuration it appears to be detecting an install location, but I don't know how to get it to work.
I see that there are ways to redirect make install, but my concern is that because some files were installed to a different relative path under /usr/local/ compared to what apt-get did when installing to /usr/, that I may end up with both old and new version installed which could cause my issues.  Also, I do not recall using the redirect option when configuring proftpd on the production server and see no evidence that I did in the command history.

#### Additional info on unexpected file deployment
See these folder listings to show what I am seeing (user and server name removed for security):
> ---:/usr$ ls bin/\*ftp\*
> ~~_bin/apt-ftparchive  bin/ftp_~~  bin/ftpcount  bin/ftpdctl  bin/ftptop  bin/ftpwho  ~~_bin/netkit-ftp  bin/pftp  bin/sftp_~~
> ---:/usr$ ls sbin/\*ftp\*
> **sbin/ftpasswd**  **sbin/ftpquota**  sbin/ftpscrub  sbin/ftpshut  sbin/ftpstats  sbin/in.proftpd  sbin/proftpd  ~~_sbin/proftpd-gencert_~~
> ---:/usr$ ls local/bin/\*ftp\*
> **local/bin/ftpasswd**  local/bin/ftpcount  local/bin/ftpdctl  ~~_local/bin/ftpmail_~~  **local/bin/ftpquota**  local/bin/ftptop  local/bin/ftpwho
> ---:/usr$ ls local/sbin/\*ftp\*
> local/sbin/ftpscrub  local/sbin/ftpshut  local/sbin/in.proftpd  local/sbin/proftpd

You can see that some of the files deployed to subfolders in "/user/local/" are in different subfolders compared to those deployed to "/usr/" subfolders, and some do not have a corresponding file in the other location.  e.g. "ftpasswd" & "ftpquota" are in "/usr/bin/" and "/usr/local/sbin/" (not "/usr/local/bin/").  "proftpd-gencert" is in "/usr/bin/", but not in a local folder (although my compiled version is not including sftp or ftps).  Note that I bolded those that are moved and struckthrough those that do not exist in the other base location.  I wish I could apply colours instead to emphasize this, alas I am limited to what markdown can do.

It is possible that I am confusing files that are not part of the proftpd package with those that are, but I am unsure and don't want to cause our production systems any issues.  Perhaps you can tell I am a bit of a newbie to Linux systems. 🙂

I hope someone can give me some insight on this because I need to eliminate the issues that were discovered with our live ftp soon. I apologize if I am not coming off well. This would probably be better on the forums, but unfortunately it cannot accept new users.

### ProFTPD Version and Configuration
As I mentioned above, I started with version 1.3.6c and am attempting to replace it with a compiled version of 1.3.7a.
The virtual machine was created using Windows 10 Hyper-V's quick create for Ubuntu 20.04.
Also, I configured it using the following command:
sudo ./configure --enable-dso --enable-ctrls --with-modules=mod_ident:mod_rewrite

I can provide the output of the configure command if you believe it is helpful, or any other info.

### Epilogue
I appreciate any advice that can be provided on this.
Note that I am quite willing to provide any additional information on this.

Thank you.

**Comment by mdell-seradex** [comment]
I hope someone can give me some insight on this because I need to eliminate the issues that were discovered with our live ftp soon.  I apologize if I am not coming off well.  This would probably be better on the forums, but unfortunately it cannot accept new users.

Thank you

**Comment by Castaglia** [comment]
Can you get the output from running `proftpd -V` on the apt-installed ProFTPD?  That output is designed for scenarios like this.  That is, `proftpd -V` should tell you the `configure` command used to build it; then you can copy/paste that `configure` command for the source code, to compile it the same way.

**Comment by mdell-seradex** [comment]
Wow.  I ran proftpd -V and this was the relevant output:
>  Built With:
>    configure  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--prefix=/usr' '--with-includes=/usr/include/postgresql:/usr/include/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--enable-redis' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab: mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security'

That is a ton of information and feels a bit overwhelming.  Most likely way more than I need.  Other information by running "proftpd -V" was also provided, but is probably more information than would be required for this.
I'm not really sure what to pull out of this to get things to line up.  I am surprised that when I compiled it in the production environment it seemed to just work.
I definitely do not need any of the SQL stuff.  I don't really have any significant logging setup either.  I think it may be none at this time.

Currently this just needs to be a pretty simple FTP server similar to the configuration I was using, however, if there is anything you recommend that I should include, for things like security, I am quite willing to do so.  Note that the configuration for the FTP server used in production is blocking directory listings and uploading of files so that helps with security.
I saw '--prefix=/usr' as part of that configuration, so I guess that begins to explain the differences.  It seems like many other folders are specified as well.
I am unsure if '{prefix}' is valid to use in an actual configuration as I have not yet seen that anywhere but here, but there is so much to read and no time to try to do that.

My next step is to try to implement your mod_case in the configuration.  I am surprised that it is not included with the official proftpd "Contrib modules" since you seem to be a main contributor here.  Hopefully I can use that extra module without making additional changes. 😃

I appreciate the input so far and would greatly appreciate any additional feedback you can provide.

Thank you very much.

Mark

**Comment by mdell-seradex** [comment]
@Castaglia I want to thank you for your help so far and wonder if you could offer assistance in pulling the essentials out of that configuration.  Also, I am a bit unsure about the format displayed.  My guess is that the single quotation marks are included to help separate things out, but it just seems to add more confusion to me.  In addition, as I sort of mentioned in my previous post, do I have to replace that "{prefix}" string with a relevant piece of info, or does the configuration recognize this and replace with an appropriate value, whatever that may be I don't know, as needed?

Thank you

**Comment by mdell-seradex** [comment]
I was thinking about this further and thought the following may provide more information and could affect any recommendations you have:
> $ proftpd -l
> Compiled-in modules:
>   mod_core.c  mod_xfer.c  mod_rlimit.c  mod_auth_unix.c  mod_auth_file.c  mod_auth.c  mod_ls.c  mod_log.c  mod_site.c  mod_delay.c  mod_facts.c  mod_dso.c  mod_ident.c  mod_auth_pam.c  mod_cap.c  mod_ctrls.c  mod_lang.c
That seems like far fewer modules than "proftpd -V" includes, and is not from the compiled version, but from the apt-get installed version.

proftp.conf file:
```
# Includes DSO modules
Include /etc/proftpd/modules.conf

# If set on you can experience a longer connection delay in many cases.
IdentLookups                   off

ServerType                     standalone
DeferWelcome                   off

MultilineRFC2228               on
DefaultServer                  on
ShowSymlinks                   on

TimeoutNoTransfer              600
TimeoutStalled                 600
TimeoutIdle                    1200

DisplayLogin                   welcome.msg
DisplayChdir                   .message true
ListOptions                    "-l"

DenyFilter                     \*.*/

# Use this to jail all users in their homes 
DefaultRoot                    ~

# Port 21 is the standard FTP port.  Port 0 disables the generic setup
Port                           0

# Ensures that a TCP listening socket is created and bound to a specific IP address for the main "server config" server and all configured virtual servers instead of a wildcard binding to a specific port.
SocketBindTight                on

# To prevent DoS attacks, set the maximum number of child processes to 30.  If you need to allow more than 30 concurrent connection at once, simply increase this value.  Note that this ONLY works in standalone mode, in inetd mode you should use an inetd server that allows you to limit maximum number of processes per service (such as xinetd)
MaxInstances                   30

# Set the user and group that the server normally runs at.
User                           proftpd
Group                          nogroup

# Umask 022 is a good standard umask to prevent new files and dirs (second parm) from being group and world writable.
Umask                          022  022
# Normally, we want files to be overwriteable.
AllowOverwrite                 on

TransferLog /var/log/proftpd/xferlog
SystemLog   /var/log/proftpd/proftpd.log

# In order to keep log file dates consistent after chroot, use timezone info from /etc/localtime.  If this is not set, and proftpd is configured to chroot (e.g. DefaultRoot or <Anonymous>), it will use the non-daylight savings timezone regardless of whether DST is in effect.
SetEnv TZ :/etc/localtime

<IfModule mod_quotatab.c>
QuotaEngine off
</IfModule>

<IfModule mod_ratio.c>
Ratios off
</IfModule>

# Delay engine reduces impact of the so-called Timing Attack described in http://www.securityfocus.com/bid/11430/discuss  It is on by default. 
<IfModule mod_delay.c>
DelayEngine on
</IfModule>

<IfModule mod_ctrls.c>
ControlsEngine        off
ControlsMaxClients    2
ControlsLog           /var/log/proftpd/controls.log
ControlsInterval      5
ControlsSocket        /var/run/proftpd/proftpd.sock
</IfModule>

<IfModule mod_ctrls_admin.c>
AdminControlsEngine off
</IfModule>

# Useful to keep VirtualHost/VirtualRoot directives separated
Include /etc/proftpd/virtuals.conf

# Include other custom configuration files !! Please note, that this statement will read /all/ file from this subdir, i.e. backup files created by your editor, too !!!  Eventually create file patterns like this: /etc/proftpd/conf.d/*.conf
# NOTE that the "conf.d" folder is currently empty.
Include /etc/proftpd/conf.d/
```

virtuals.conf file:
```
<VirtualHost 172.19.143.198>
  ServerName              "Custom name"
  Port                    21
  TransferLog             /var/log/proftpd/xfer/support.seradex.com.log
  MaxLoginAttempts        1
  RequireValidShell       no
  DefaultRoot             /ftp
  AllowOverwrite          no
  Protocols               ftp
  # Use only AuthUserFiles when authenticating, and not the system's /etc/passwd
  AuthOrder               mod_auth_file.c
  AuthGroupFile           /etc/proftpd/ftpd.group
  AuthUserFile            /etc/proftpd/ftpd.passwd

  <Directory ~>
    HideNoAccess on
    <Limit WRITE CDUP MDTM MLSD MLST NLST PWD RNFR XCUP XCWD XPWD MFMT SITE_CHGRP SITE_CHMOD SITE_MKDIR SITE_RMDIR SITE_SYMLINK SITE_UTIME>
      DenyAll
    </Limit>
  </Directory>

  <IfModule mod_rewrite.c>
    RewriteEngine on

    # Define a map that uses the internal "replaceall" function
    RewriteMap replace int:replaceall

    # We only want to use this rule on RETR and SIZE commands
    RewriteCondition %m RETR|SIZE|LIST

    # Apply the map to the command parameters
    RewriteRule (.*) "${replace:!$1!\\!/}"
  </IfModule>
</VirtualHost>

```

**Comment by Castaglia** [comment]
My recommendation is that you focus on getting the source code compiled _in the same way as the `apt` package_ first.  That way, you can use the new build, have it work with the old configuration, and get a working ProFTPD installation sooner rather than later.

Only once _that_ has been done, would it be time to look at what bits you might not need.  If you do things the other way around, you may find yourself spending more time on getting a working configuration than you want.

**Comment by mdell-seradex** [comment]
@Castaglia, I completely understand what you are saying.  The odd thing is that I am 99% certain compiling it on our production machine, which started as an actual copy of this one, worked.
One thing that I had found was that certain modules were not detected and compiled in because of what is currently installed on the Ubuntu machine.  I found that I had to install a version of libcap to get capabilities to get detected and included.
Another thing you can see is that "mod_rewrite" is part of the configuration, but is not listed as a compiled in module by using "proftpd -l", and that is a module we use on our production server.
This discrepancy seems very odd.

**Comment by Castaglia** [comment]
The `proftpd -l` output lists only _static_ modules; from the `proftpd -V` output you provided, it looks like the `mod_rewrite` module is compiled as a _shared_ (DSO) module, which means it would be dynamically loaded into your `proftpd` executable, using a configuration like:
```
<IfModule mod_dso.c>
  LoadModule mod_rewrite.c
</IfModule>
```

**Comment by mdell-seradex** [comment]
@Castaglia  Thank you.  That makes so much more sense now.
It sounds like my attempt to recompile the newer version wasn't really required, but of course I had no clue.

For configuring the new version, am I right that I would use that same text that "proftpd -V" gave me, but remove the single quote marks (aka apostrophes)?
So I would take this `configure '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' ...` and execute it as this `./configure --build=x86_64-linux-gnu --includedir=${prefix}/include --mandir=${prefix}/share/man --infodir=${prefix}/share/info ...`?  (I left a lot out for brevity, but I essentially mean that entire string.)  Or is removal of the single quote marks (apostrophes) not required?  My guess and hope is that if I need to change things then it isn't much.

And I take it that the use of "{prefix}" in there is nothing I need to change to something else?

I hope I will not need to use apt-get to install other things to allow it to "configure" and "make" correctly.
I was unsure when I did the previous attempt about the correct version of libcap to install.  I ran this to do it `sudo apt-get install libcap2-bin`

**Comment by mdell-seradex** [comment]
Oh no 😦 @Castaglia.

I just noticed that `sudo apt install proftpd-dev`  was run on the production environment at some point while I was trying to compile things for version 1.3.7a.  Like I said, I am a bit of a newbie.
So the settings are very different there.  That probably explains why the build worked.
It says that it was `  Built With: configure  '--enable-dso' '--enable-ctrls' '--with-modules=mod_ident:mod_rewrite'`, but it also says this is the `Configuration File: /usr/local/etc/proftpd.conf`, which is completely different from my configuration, and I know my configuration is functioning because when I access the FTP server it is preventing me from browsing.  I can only download files if I know the name and location, and I cannot upload anything.
The configuration file `proftpd.conf` specified by running `proftpd -V` on the production server does not limit directory browsing like mine does.  That files only has the following limits.
```<Limit SITE_CHMOD>
  DenyAll
</Limit>
...
<Anonymous>
...
  # Limit WRITE everywhere in the anonymous chroot
  <Limit WRITE>
    DenyAll
  </Limit>
</Anonymous>
```
It was originally running the version installed using `sudo apt-get install proftpd`, but now seems to be overwritten with the dev version.  My guess is that is not really a good thing.

Would you recommend uninstalling and starting over again, keeping my configuration files?  Not sure how to currently, but I suspect it would be easy to look up.  The original reason to compile it was to get the `mod_rewrite` functionality, but it seems that I now need that `mod_case` ability which is what started this entire issue (question).

I am sorry that I keep bugging you 🥺.  I wish the forums were working so I could just post there.
I just don't want to screw this up, which it seems I already have slightly, since an important part of our business is making use of this.  🤦‍♂️

**Comment by Castaglia** [comment]
Not sure if you ever use tools like Docker or Vagrant; I find these useful, for spinning up "disposable" development environments, of the same Linux distribution as my production environment, and then doing testing/development in them -- knowing that then, I won't break/affect production in any way.

For example:
```
$ docker run --rm=true -it ubuntu:20.04 bash
root@a89451e244f4:/# apt-get update
root@a89451e244f4:/# apt-get install -y proftpd
root@a89451e244f4:/# proftpd -V
Compile-time Settings:
  Version: 1.3.6c (maint)
  Platform: LINUX [Linux 4.9.49-moby x86_64]
  Built: Thu Feb 27 2020 19:34:56 UTC
  Built With:
    configure  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' ...

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd
...
```
You were asking about that `{prefix}` variable mentioned in the above output; that is probably a variable in Debian's build system.  And using the "Files" section of the output, we can be pretty sure that `{prefix}` is `/usr` in this case.

If you haven't done so yet, I highly recommend checking in your config files into a source code control system (git, subversion, Perforce; doesn't really matter).  That way, you have a backup of your config files -- and you can track the history of changes over time.

**Comment by mdell-seradex** [comment]
Right now this Linux Ubuntu system is the only one I am managing, and I don't see that changing soon.
It is odd that it would be a Debian variable since I thought the package I installed would be specific to Ubuntu.

Regarding my screw up, I don't know for sure, especially since I a new to Linux, but I guess uninstalling and starting over would give me a clean starting place.

Unfortunately I do not really have time to learn Docker or Vagrant at this time.
I am taking a local copy of my config files, and have no intent to make any more changes at this time once the issue that was pointed out to me has been addressed, for which, hopefully `mod_case` will address.

I am a bit afraid because of how I screwed up, but I guess I will just have to try.

**Comment by mdell-seradex** [comment]
I tried to run `sudo make clean` followed by the `sudo ./configure` with the parameters specified by `proftpd -V`.  I tried removing the single quotes from the configuration, but that only make it fail sooner.  It says it cannot recognize parameter O2 (or something like that).
After apparently successfully configuring I ran `sudo make` and I then got the following error (with the associated compile section)
```
libtool: compile:  gcc -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_CONFIG_H -DLINUX -I.. -I../include -I../include -I/usr/include/mysql -I/usr/include/postgresql -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -c pr-syslog.c  -fPIC -DPIC -o .libs/pr-syslog.o
In file included from ../include/conf.h:75,
                 from pr-syslog.c:30:
../include/regexp.h:36:11: fatal error: pcre.h: No such file or directory
   36 | # include <pcre.h>
      |           ^~~~~~~~
compilation terminated.
make[1]: *** [Makefile:25: pr-syslog.o] Error 1
make[1]: Leaving directory '/usr/src/proftpd-1.3.7a/lib'
make: *** [Makefile:51: lib] Error 2
```
Obviously I am missing something, but I know not what.

**Comment by mdell-seradex** [comment]
I tried to run `sudo make clean` followed by the `sudo ./configure` with the parameters specified by `proftpd -V`.  I tried removing the single quotes from the configuration, but that only make it fail sooner.  It says it cannot recognize parameter O2 (or something like that).
After apparently successfully configuring I ran `sudo make` and I then got the following error (with the associated compile section)
```
libtool: compile:  gcc -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_CONFIG_H -DLINUX -I.. -I../include -I../include -I/usr/include/mysql -I/usr/include/postgresql -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -c pr-syslog.c  -fPIC -DPIC -o .libs/pr-syslog.o
In file included from ../include/conf.h:75,
                 from pr-syslog.c:30:
../include/regexp.h:36:11: fatal error: pcre.h: No such file or directory
   36 | # include <pcre.h>
      |           ^~~~~~~~
compilation terminated.
make[1]: *** [Makefile:25: pr-syslog.o] Error 1
make[1]: Leaving directory '/usr/src/proftpd-1.3.7a/lib'
make: *** [Makefile:51: lib] Error 2
```
Obviously I am missing something, but I know not what.

**Comment by mdell-seradex** [comment]
Hi @Castaglia.
Hope everything is well with you.
I keep accidentally hitting Ctrl-Enter to try to create a new line and it saves instead.  🤦‍♂️

I tried to run `sudo make clean` followed by the `sudo ./configure` with the parameters specified by `proftpd -V`.  I tried removing the single quotes from the configuration, but that only make it fail sooner.  It says it cannot recognize parameter O2 (or something like that).
After apparently successfully configuring I ran `sudo make` and I then got the following error (with the associated compile section)
```
libtool: compile:  gcc -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_CONFIG_H -DLINUX -I.. -I../include -I../include -I/usr/include/mysql -I/usr/include/postgresql -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -c pr-syslog.c  -fPIC -DPIC -o .libs/pr-syslog.o
In file included from ../include/conf.h:75,
                 from pr-syslog.c:30:
../include/regexp.h:36:11: fatal error: pcre.h: No such file or directory
   36 | # include <pcre.h>
      |           ^~~~~~~~
compilation terminated.
make[1]: *** [Makefile:25: pr-syslog.o] Error 1
make[1]: Leaving directory '/usr/src/proftpd-1.3.7a/lib'
make: *** [Makefile:51: lib] Error 2
```
Obviously I am missing something, but I know not what.

I am not sure what the pcre.h header file is for.

**Comment by mdell-seradex** [comment]
Hi @Castaglia.
Hope everything is well with you.
I keep accidentally hitting Ctrl-Enter to try to create a new line and it saves instead.  🤦‍♂️

I tried to run `sudo make clean` followed by the `sudo ./configure` with the parameters specified by `proftpd -V`.  I tried removing the single quotes from the configuration, but that only make it fail sooner.  It says it cannot recognize parameter O2 (or something like that).
After apparently successfully configuring I ran `sudo make` and I then got the following error (with the associated compile section)
```
libtool: compile:  gcc -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_CONFIG_H -DLINUX -I.. -I../include -I../include -I/usr/include/mysql -I/usr/include/postgresql -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -c pr-syslog.c  -fPIC -DPIC -o .libs/pr-syslog.o
In file included from ../include/conf.h:75,
                 from pr-syslog.c:30:
../include/regexp.h:36:11: fatal error: pcre.h: No such file or directory
   36 | # include <pcre.h>
      |           ^~~~~~~~
compilation terminated.
make[1]: *** [Makefile:25: pr-syslog.o] Error 1
make[1]: Leaving directory '/usr/src/proftpd-1.3.7a/lib'
make: *** [Makefile:51: lib] Error 2
```
Obviously I am missing something, but I know not what.

I am not sure what the pcre.h header file is for or where it is supposed to come from.

**Comment by mdell-seradex** [comment]
Hi @Castaglia.
Hope everything is well with you.
I keep accidentally hitting Ctrl-Enter to try to create a new line and it saves instead.  🤦‍♂️

I tried to run `sudo make clean` followed by the `sudo ./configure` with the parameters specified by `proftpd -V`.  I tried removing the single quotes from the configuration, but that only make it fail sooner.  It says it cannot recognize parameter O2 (or something like that).
After apparently successfully configuring I ran `sudo make` and I then got the following error (with the associated compile section)
```
libtool: compile:  gcc -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_CONFIG_H -DLINUX -I.. -I../include -I../include -I/usr/include/mysql -I/usr/include/postgresql -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -c pr-syslog.c  -fPIC -DPIC -o .libs/pr-syslog.o
In file included from ../include/conf.h:75,
                 from pr-syslog.c:30:
../include/regexp.h:36:11: fatal error: pcre.h: No such file or directory
   36 | # include <pcre.h>
      |           ^~~~~~~~
compilation terminated.
make[1]: *** [Makefile:25: pr-syslog.o] Error 1
make[1]: Leaving directory '/usr/src/proftpd-1.3.7a/lib'
make: *** [Makefile:51: lib] Error 2
```
Obviously I am missing something, but I know not what.

I am not sure what the pcre.h header file is for or where it is supposed to come from.
I am running the commands directly in the `proftpd-1.3.7a` folder.
You can see above that I am keeping it in `/usr/src/`.  I didn't think the actual location would truly make a difference, as long as it is not too long.
I really wasn't sure where I should put it and usually prefer not to keep things in user folders that actually aren't user specific.
Yes, I am doing this from my user account, but not for my user.  This way if my profile is dropped, these files are not lost.

Do you have any idea what I am doing wrong or I am missing?
Perhaps this corresponds to things I don't need so I can exclude them?  The build on did on the production server, although was wrong, seems to be working as desired.
I will do a new proper test once I get this working to ensure all that we need is correctly working.

**Comment by mdell-seradex** [comment]
Hi @Castaglia.
Hope everything is well with you.
I keep accidentally hitting Ctrl-Enter to try to create a new line and it saves instead.  🤦‍♂️

I tried to run `sudo make clean` followed by the `sudo ./configure` with the parameters specified by `proftpd -V`.  I tried removing the single quotes from the configuration, but that only make it fail sooner.  It says it cannot recognize parameter O2 (or something like that).
After apparently successfully configuring I ran `sudo make` and I then got the following error (with the associated compile section)
```
libtool: compile:  gcc -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_CONFIG_H -DLINUX -I.. -I../include -I../include -I/usr/include/mysql -I/usr/include/postgresql -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -c pr-syslog.c  -fPIC -DPIC -o .libs/pr-syslog.o
In file included from ../include/conf.h:75,
                 from pr-syslog.c:30:
../include/regexp.h:36:11: fatal error: pcre.h: No such file or directory
   36 | # include <pcre.h>
      |           ^~~~~~~~
compilation terminated.
make[1]: *** [Makefile:25: pr-syslog.o] Error 1
make[1]: Leaving directory '/usr/src/proftpd-1.3.7a/lib'
make: *** [Makefile:51: lib] Error 2
```
Obviously I am missing something, but I know not what.

I am not sure what the pcre.h header file is for or where it is supposed to come from.
I am running the commands directly in the `proftpd-1.3.7a` folder.
You can see above that I am keeping it in `/usr/src/`.  I didn't think the actual location would truly make a difference, as long as it is not too long.
I really wasn't sure where I should put it and usually prefer not to keep things in user folders that actually aren't user specific.
Yes, I am doing this from my user account, but not for my user.  This way if my profile is dropped, these files are not lost.

Do you have any idea what I am doing wrong or I am missing?
Perhaps this corresponds to things I don't need so I can exclude them?  The build on did on the production server, although was wrong, seems to be working as desired.
I will do a new proper test once I get the `make` and `make install` working to ensure all that we need from proftpd is working correctly.

**Comment by mdell-seradex** [comment]
Hi @Castaglia.
Hope everything is well with you.
I keep accidentally hitting Ctrl-Enter to try to create a new line and it saves instead.  🤦‍♂️

I tried to run `sudo make clean` followed by the `sudo ./configure` with the parameters specified by `proftpd -V`.  I tried removing the single quotes from the configuration, but that only make it fail sooner.  It says it cannot recognize parameter O2 (or something like that).
After apparently successfully configuring I ran `sudo make` and I then got the following error (with the associated compile section)
```
libtool: compile:  gcc -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_CONFIG_H -DLINUX -I.. -I../include -I../include -I/usr/include/mysql -I/usr/include/postgresql -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration -c pr-syslog.c  -fPIC -DPIC -o .libs/pr-syslog.o
In file included from ../include/conf.h:75,
                 from pr-syslog.c:30:
../include/regexp.h:36:11: fatal error: pcre.h: No such file or directory
   36 | # include <pcre.h>
      |           ^~~~~~~~
compilation terminated.
make[1]: *** [Makefile:25: pr-syslog.o] Error 1
make[1]: Leaving directory '/usr/src/proftpd-1.3.7a/lib'
make: *** [Makefile:51: lib] Error 2
```
Obviously I am missing something, but I know not what.

I am not sure what the pcre.h header file is for or where it is supposed to come from.
I am running the commands directly in the `proftpd-1.3.7a` folder.
You can see above that I am keeping it in `/usr/src/`.  I didn't think the actual location would truly make a difference, as long as it is not too long.
I really wasn't sure where I should put it and usually prefer not to keep things in user folders that actually aren't user specific.
Yes, I am doing this from my user account, but not for my user.  This way if my profile is dropped, these files are not lost.

Do you have any idea what I am doing wrong or I am missing?

Perhaps this corresponds to things I don't need so I can exclude them?  The build on did on the production server, although was wrong, seems to be working as desired.
I will do a new proper test once I get the `make` and `make install` working to ensure all that we need from proftpd is working correctly.

Thank you for all your help.
It is appreciated.

Mark

**Comment by Castaglia** [comment]
Since your ultimate goal is to add the `mod_case` module, you can use ProFTPD's `prxs` (ProFTPD Extensions) tool, to build `mod_case` _without rebuilding ProFTPD_, and loading it as a shared module, using _e.g._:
```
# The `proftpd-dev` package provides the necessary `prxs` tool
$ apt-get install -y proftpd proftpd-dev

# Download mod_case from GitHub
$ git clone https://github.com/Castaglia/proftpd-mod_case.git mod_case
$ cd mod_case/

# Build and install mod_case as a shared module
$ prxs -c -i -d mod_case.c
```
Then, in your ProFTPD config, near the start, you tell ProFTPD to dynamically load it:
```
<IfModule mod_dso.c>
  LoadModule mod_case.c
</IfModule>
```
After that, you should be able to configure/use the `mod_case` module as needed.

**Comment by mdell-seradex** [comment]
Thank you so much for that @Castaglia.
The reason why I built version 1.3.7a in the first place was because I needed mod_rewrite and did not believe that it was included.
I now know that my belief was incorrect and I just had to get it to load.  I still have to try that.
Hopefully the two modules will work well together.
I hope to test this later today and with luck, all my issues will be dealt with!

I do wonder if it would be beneficial to be on a newer version, but since as of the last time I checked, it is not available via apt-get, I will leave that for the time being.

Thank you again for your help.
I will let you know how things work out.

**Comment by mdell-seradex** [comment]
@Castaglia, I had to do a lot of clean up, deleting files in different folder under `/usr/local`, to get it to install properly.
It is now installed in my test server running 1.3.6a and proftpd starts successfully.
Before the clean up, it seemed to install to the wrong location since proftpd would not start with the following in the proftpd.conf file.
```
<IfModule mod_dso.c>
  LoadModule mod_case.c
</IfModule>
```
That said, `mod_case.c` works well as long as it is not a folder in the middle of the provided path that has the wrong case.
I am going to attempt to modify it to address this.  🤞

**Comment by mdell-seradex** [comment]
@Castaglia, I had to do a lot of clean up, deleting files in different folders under `/usr/local`, to get it to install properly.
It is now installed in my test server running 1.3.6a and proftpd starts successfully.
Before the clean up, it seemed to install to the wrong location since proftpd would not start with the following in the proftpd.conf file.
```
<IfModule mod_dso.c>
  LoadModule mod_case.c
</IfModule>
```
That said, `mod_case.c` works well as long as it is not a folder in the middle of the provided path that has the wrong case.
e.g.   Correct path: `/App12/modules/test.c`  incorrect path: `/App12/MODULES/test.c`
I am going to attempt to modify it to address this.  🤞

**Comment by mdell-seradex** [comment]
I have modified mod_case.c for my requirements.
I even created my first

**Comment by mdell-seradex** [comment]
Hi @Castaglia!
I have modified mod_case.c for my requirements.
I even created my first GitHub repository (a fork of yours) so I could publish my changes and submit a pull request to you (which I have done).
I also added you as a contributor on my fork.
As I said in the pull request, I did not change all of the methods that call your `case_have_file` to use the replacement I created `find_path_case_insensitive` as it took longer than I hoped to get things working and I am a bit concerned that I have made an error in memory management or something else.

Note that I did test it in my development server and it is working correctly, however I do have one more test to run.
I hope that you like my changes and implement them or some similar form of them, plus whatever changes your feel are required, especially in the other two methods, for all of it to work properly.

I would greatly appreciate it if you could at least review it prior to me deploying my version to our production server.  As I said in the pull request, if there are any changes you would like me to make (preferably not too difficult), I am quite happy to do them.  Note that I rarely code in c and c++.
One quick change I can make, if you think that makes it easier for you, is to delete all the commented out code lines.  I should probably also remove the comment about the "confusing variable names" that I accidentally left in.  Sorry 😞.

Please let me know, and thank you for all of your help with this. 😄 

Mark

**Comment by mdell-seradex** [comment]
Thanks to the information provided here, I have a much better understanding of how this works and now see that I actually do not need to do this.
Hopefully, if I decide that I need to do this in the future, this will provide enough information to complete the task.

Thank you for all of your help.

**Comment by mdell-seradex** [comment]
I just thought I would mention that I now have deployed the latest mod_case.c and cleaned up the files deployed by make install as best as I could. Unfortunately there is no uninstall script provided with proftpd.
All seems to be working as desired. 

In doing so, I discovered that apparently I was never actually running v1.3.7a.  It figures.
Commands were simply responding as if it were true because of how another employee had modified the $PATH variable.
That, combined with other things working as desired, which I did not believe was possible with the installed 1.3.6c version, due to my limited knowledge and understanding of things.
At least my knowledge has greatly improved due to this experience.
